### PR TITLE
[In progress] Resource factory

### PR DESCRIPTION
**Pull request by nein09**
The skeleton of the resource factories. Based on the canvas-object branch.

I cut the thumbnails down to quarter-size, and the memory problems were alleviated, which makes sense.

This branch is primarily for https://github.com/sul-dlss/iiifManifestLayouts/issues/98.

### CODE REVIEW

File: src/imageResource.js [hunk 1]
```diff
@@ -80,7 +78,7 @@ ImageResource.prototype = {
             self.visible = true;
             self.status = 'shown';
             self.parent.viewer.removeHandler('tile-drawn', tileDrawnHandler);
-            self.dispatcher.emit('image-resource-tile-source-opened', { 'detail': self.tileSource });
+            self.parent.dispatcher.emit('image-resource-tile-source-opened', { 'detail': self.tileSource });
```
Comment (iangilman) [inline src/imageResource.js]: You do have a `self.dispatcher`, so you don't need to use `self.parent.dispatcher`. Of course yours is just copied from theirs, but I think that's better to do it once than throughout the code...that way if we change to a different dispatcher it's an easier change.

File: src/imageResourceFactory.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+require('openseadragon');
+var ImageResource = require('./ImageResource');
+
+var _buildImageConfig = function(image) {
+  var _getImageTilesource = function() {
+    if(image.resource.service) {
+      return image.resource.service['@id'] + '/info.json';
+    } else {
+      return image.resource['@id'];
```
Comment (iangilman) [inline src/imageResourceFactory.js]: My note on #102 about this not being a tile source holds here as well.

File: src/imageResourceFactory.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  var resourceType = image.resource['@type']; // can be oa:Choice, oa:SpecificResource, or dctypes:Image
+  var config = {};
+
+  switch(resourceType) {
+    case 'dctypes:Image':
+      config = _buildImageConfig(image);
+      break;
+    case 'oa:Choice':
+      console.log('oa:Choice!', image);
+      config = _buildChoiceResource(image);
```
Comment (iangilman) [inline src/imageResourceFactory.js]: Shouldn't this be called _buildChoiceConfig for consistency?

File: src/thumbnailFactory.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+var ImageResource = require('./ImageResource');
+
+var _getThumbService = function(image, width) {
```
Comment (iangilman) [inline src/thumbnailFactory.js]: I guess my comment about where this should go in #102 is moot, but the naming comment still holds.

**Comment by iangilman** [comment]
Seems like the right direction to me!

File: src/thumbnailFactory.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  if(image.resource.service) {
+    return image.resource.service['@id'] + '/full/' + Math.ceil(width / 4) + ',/0/default.jpg';
+  } else {
+    return image.resource['@id'];
```
Comment (nein09) [inline src/thumbnailFactory.js:9]: Maybe what I could do here, to address one of your comments in #102 is if there's only one image in the canvas and it's static, I could build a config of imageType 'main' and check for that in the CanvasObject when it gets the thumbnail back. Then the CanvasObject could just not bother opening the real tilesource.

Comment (iangilman) [inline src/thumbnailFactory.js:9]: Something like that sounds reasonable.

**Comment by nein09** [comment]
Going to reopen this PR on master in a bit.

**Comment by aeschylus** [comment]
Okay. I'll have a look then. Would it help if I started setting up the example page?

**Comment by nein09** [comment]
If you're inclined to do that, it'd be great- at the moment, I'm working my way through the current example page to see what I'm not handling for the initial opening of the manifest. I just got through number 29.

**Pull request closed without merging**